Throw a useful error if option isn't set

```python
assert s.show_options('default-shell', g=True)
```

```python
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../libtmux/libtmux/session.py:374: in show_options
    return self.show_option(option, g=g)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Session($1 libtmux_XxTayh), option = [], g = True

    def show_option(self, option, g=False):
        """Return a list of options for the window.

            :todo: test and return True/False for on/off string

            :param option: option to return.
            :type option: string
            :rtype: string, int or bool

            """

        tmux_args = tuple()

        if g:
            tmux_args += ('-g',)

        option = self.cmd(
            'show-options', option, *tmux_args
        ).stdout
>       option = [tuple(item.split(' ')) for item in option][0]
E       IndexError: list index out of range
```

Have a ``raise=False`` argument. If not raising, return ``None``.

--- a/CHANGES
+++ b/CHANGES
@@ -4,6 +4,9 @@ Changelog
 
 Here you can find the recent changes to libtmux
 
+- :feature:`8` new exception ``UnknownOption``
+- :feature:`8` return ``None`` for options that are valid tmux options,
+  but unset at that scope.
 - :support:`6` major documentation overhaul
 
 - :release:`0.4.1 <2016-05-23>`

--- a/doc/about.rst
+++ b/doc/about.rst
@@ -13,7 +13,8 @@ libtmux is an `abstraction layer`_ for tmux.
 
 It builds upon the concept of targets ``-t``, to direct commands against
 individual session, windows and panes and ``FORMATS``, template variables 
-exposed by tmux to describe their properties.
+exposed by tmux to describe their properties. Think of ``-t`` analagously
+to `scope`_.
 
 :class:`common.TmuxRelationalObject` acts as a container to connect the
 relations of :class:`Server`, :class:`Session`, :class:`Window` and
@@ -101,3 +102,4 @@ Reference
 - tmux source code http://sourceforge.net/p/tmux/tmux-code/ci/master/tree/
 
 .. _abstraction layer: http://en.wikipedia.org/wiki/Abstraction_layer
+.. _scope: https://en.wikipedia.org/wiki/Variable_(computer_science)#Scope_and_extent

--- a/doc/api.rst
+++ b/doc/api.rst
@@ -73,6 +73,8 @@ Exceptions
 
 .. autoexception:: libtmux.exc.TmuxSessionExists
 
+.. autoexception:: libtmux.exc.UnknownOption
+
 Test tools
 ----------
 

--- a/libtmux/exc.py
+++ b/libtmux/exc.py
@@ -19,3 +19,10 @@ class TmuxSessionExists(LibTmuxException):
     """Session does not exist in the server."""
 
     pass
+
+
+class UnknownOption(LibTmuxException):
+
+    """Option unknown to tmux show-option(s) or show-window-option(s)."""
+
+    pass

--- a/libtmux/session.py
+++ b/libtmux/session.py
@@ -43,14 +43,13 @@ class Session(
         self.server._update_windows()
 
     @property
-    def _info(self, *args):
+    def _info(self):
 
         attrs = {
             'session_id': str(self._session_id)
         }
 
-        # from https://github.com/serkanyersen/underscore.py
-        def by(val, *args):
+        def by(val):
             for key, value in attrs.items():
                 try:
                     if attrs[key] != val[key]:
@@ -324,7 +323,7 @@ class Session(
 
         return self.attached_window.attached_pane
 
-    def set_option(self, option, value):
+    def set_option(self, option, value, g=False):
         """Set option ``$ tmux set-option <option> <value>``.
 
         todo: needs tests
@@ -334,6 +333,8 @@ class Session(
         :param value: window value. True/False will turn in 'on' and 'off'. You
             can also enter 'on' or 'off' directly.
         :type value: bool
+        :param global: check for option globally across all servers (-g)
+        :type global: bool
 
         """
 
@@ -342,14 +343,22 @@ class Session(
         elif isinstance(value, bool) and not value:
             value = 'off'
 
+        tmux_args = tuple()
+
+        if g:
+            tmux_args += ('-g',)
+
+        tmux_args += (option, value,)
+
         proc = self.cmd(
-            'set-option', option, value
+            'set-option', *tmux_args
         )
 
-        if proc.stderr:
-            if isinstance(proc.stderr, list) and len(proc.stderr) == int(1):
-                proc.stderr = proc.stderr[0]
-            raise ValueError('tmux set-option stderr: %s' % proc.stderr)
+        if isinstance(proc.stderr, list) and len(proc.stderr):
+            error = proc.stderr[0]
+            if 'unknown option' in error:
+                raise exc.UnknownOption(error)
+            raise ValueError('tmux set-option stderr: %s' % error)
 
     def show_options(self, option=None, g=False):
         """Return a dict of options for the window.
@@ -359,7 +368,7 @@ class Session(
 
         :param option: optional. show a single option.
         :type option: string
-        :param g: Pass ``-g`` flag for global variable
+        :param g: Pass ``-g`` flag for global variable (server-wide)
         :type g: bool
         :rtype: :py:obj:`dict`
 
@@ -395,6 +404,8 @@ class Session(
 
         :param option: option to return.
         :type option: string
+        :param global: check for option globally across all servers (-g)
+        :type global: bool
         :rtype: string, int or bool
 
         """
@@ -404,15 +415,28 @@ class Session(
         if g:
             tmux_args += ('-g',)
 
-        window_option = self.cmd(
-            'show-options', option, *tmux_args
-        ).stdout
-        window_option = [tuple(item.split(' ')) for item in window_option][0]
+        tmux_args += (option,)
 
-        if window_option[1].isdigit():
-            window_option = (window_option[0], int(window_option[1]))
+        cmd = self.cmd(
+            'show-options', *tmux_args
+        )
 
-        return window_option[1]
+        if isinstance(cmd.stderr, list) and len(cmd.stderr):
+            error = cmd.stderr[0]
+            if 'unknown option' in error:
+                raise exc.UnknownOption(error)
+            else:
+                raise exc.LibTmuxException(error)
+
+        if not len(cmd.stdout):
+            return None
+
+        option = [item.split(' ') for item in cmd.stdout][0]
+
+        if option[1].isdigit():
+            option = (option[0], int(option[1]))
+
+        return option[1]
 
     def __repr__(self):
         return "%s(%s %s)" % (

--- a/libtmux/window.py
+++ b/libtmux/window.py
@@ -142,16 +142,18 @@ class Window(TmuxMappingObject, TmuxRelationalObject):
         elif isinstance(value, bool) and not value:
             value = 'off'
 
-        process = self.cmd(
+        cmd = self.cmd(
             'set-window-option',
             '-t%s:%s' % (self.get('session_id'), self.index),
             # '-t%s' % self.id,
             option, value
         )
 
-        if process.stderr:
-            if isinstance(process.stderr, list) and len(process.stderr):
-                process.stderr = process.stderr[0]
+        if isinstance(cmd.stderr, list) and len(cmd.stderr):
+            error = cmd.stderr[0]
+            if 'unknown option' in error:
+                raise exc.UnknownOption(error)
+
             raise ValueError(
                 'tmux set-window-option -t%s:%s %s %s\n' % (
                     self.get('session_id'),
@@ -159,7 +161,7 @@ class Window(TmuxMappingObject, TmuxRelationalObject):
                     option,
                     value
                 ) +
-                process.stderr
+                cmd.stderr
             )
 
     def show_window_options(self, option=None, g=False):
@@ -185,13 +187,13 @@ class Window(TmuxMappingObject, TmuxRelationalObject):
             return self.show_window_option(option, g=g)
         else:
             tmux_args += ('show-window-options',)
-            window_options = self.cmd(
+            cmd = self.cmd(
                 *tmux_args
             ).stdout
 
-        window_options = [tuple(item.split(' ')) for item in window_options]
+        cmd = [tuple(item.split(' ')) for item in cmd]
 
-        window_options = dict(window_options)
+        window_options = dict(cmd)
 
         for key, value in window_options.items():
             if value.isdigit():
@@ -219,20 +221,22 @@ class Window(TmuxMappingObject, TmuxRelationalObject):
 
         tmux_args += (option,)
 
-        window_option = self.cmd(
+        cmd = self.cmd(
             'show-window-options', *tmux_args
-        ).stdout
+        )
 
-        if window_option:
-            window_option = [tuple(item.split(' '))
-                             for item in window_option][0]
-        else:
+        if len(cmd.stderr) and 'unknown option' in cmd.stderr[0]:
+            raise exc.UnknownOption(cmd.stderr[0])
+
+        if not len(cmd.stdout):
             return None
 
-        if window_option[1].isdigit():
-            window_option = (window_option[0], int(window_option[1]))
+        option = [item.split(' ') for item in cmd.stdout][0]
 
-        return window_option[1]
+        if option[1].isdigit():
+            option = (option[0], int(option[1]))
+
+        return option[1]
 
     def rename_window(self, new_name):
         """Return :class:`Window` object ``$ tmux rename-window <new_name>``.

--- a/tests/test_session.py
+++ b/tests/test_session.py
@@ -8,7 +8,7 @@ import logging
 
 import pytest
 
-from libtmux import Pane, Session, Window
+from libtmux import Pane, Session, Window, exc
 from libtmux.test import TEST_SESSION_PREFIX, namer
 
 logger = logging.getLogger(__name__)
@@ -130,9 +130,19 @@ def test_set_show_option(session):
     assert session.show_option('history-limit') == 40
 
 
+def test_empty_session_option_returns_None(session):
+    assert session.show_option('default-shell') is None
+
+
+def test_show_option_unknown(session):
+    """Session.show_option raises UnknownOption for bad option key."""
+    with pytest.raises(exc.UnknownOption):
+        session.show_option('moooz')
+
+
 def test_set_option_bad(session):
-    """Session.set_option raises ValueError for bad option key."""
-    with pytest.raises(ValueError):
+    """Session.set_option raises UnknownOption for bad option key."""
+    with pytest.raises(exc.UnknownOption):
         session.set_option('afewewfew', 43)
 
 

--- a/tests/test_window.py
+++ b/tests/test_window.py
@@ -7,7 +7,7 @@ from __future__ import (absolute_import, division, print_function,
 import logging
 import pytest
 
-from libtmux import Pane, Server, Window
+from libtmux import Pane, Server, Window, exc
 
 
 logger = logging.getLogger(__name__)
@@ -169,6 +169,11 @@ def test_set_show_window_options(session):
     assert window.show_window_options()['main-pane-height'] == 40
 
 
+def test_empty_window_option_returns_None(session):
+    window = session.new_window(window_name='test_window')
+    assert window.show_window_option('default-shell') is None
+
+
 def test_show_window_option(session):
     """Set option then Window.show_window_option(key)."""
     window = session.new_window(window_name='test_window')
@@ -181,10 +186,17 @@ def test_show_window_option(session):
     assert window.show_window_option('main-pane-height') == 40
 
 
+def test_show_window_option_unknown(session):
+    """Window.show_window_option raises UnknownOption for bad option key."""
+    window = session.new_window(window_name='test_window')
+    with pytest.raises(exc.UnknownOption):
+        window.show_window_option('moooz')
+
+
 def test_set_window_option_bad(session):
     """Window.set_window_option raises ValueError for bad option key."""
 
     window = session.new_window(window_name='test_window')
 
-    with pytest.raises(ValueError):
+    with pytest.raises(exc.UnknownOption):
         window.set_window_option('afewewfew', 43)

